File: src/js/model/event.js
'use strict';

/*
 * イベント詳細 モデル
 *
 */

// API URL
var api_url = "api/event";


var m = require('../mithril');

// コメント モデル
var CommentModel = require('./comment');

// コメント モデル
var JoinModel = require('./join');



// コンストラクタ
var Model = function (data, isInitial) {
	var self = this;

	if( ! data) {
		data = {};
	}
	self.id          = m.prop(data.id);
	self.name        = m.prop(data.name        || "");
	self.place       = m.prop(data.place       || "");
	self.image       = m.prop(data.image);
	self.capacity    = m.prop(data.capacity    || "");
	self.start_date  = m.prop(data.start_date  || "");
	self.description = m.prop(data.description || "");

	// 主催者
	if(data.admin) {
		self.admin = {
			name:m.prop(data.admin.name),
		}
	} else {
		self.admin = {
			name:m.prop(""),
		};
	}

	// 場所
	if(data.place) {
		self.place = {
			name: m.prop(data.place),
		};
	}
	else {
		self.place = {
			name: m.prop(""),
		};
	}

	// 参加者一覧
	if(data.members) {
		self.members = [];
		data.members.forEach(function(member) {
			self.members.push(new JoinModel(member));
		});
	}
	
	// コメント一覧
	if(data.comments) {
		self.comments = [];
		data.comments.forEach(function(comment) {
			self.comments.push(new CommentModel(comment));
		});
	}

	self.isInitial = m.prop(true); // サーバーにレコードが存在しない
};

// サーバからJSONを読み込む
Model.read = function (id) {
	return m.request({
		method: "GET",
		url: api_url + "/" + id,
		type: Model
	})
	.then(function(model) {
		model.isInitial(false); // サーバーにレコードが存在する

		return model;
	});
};

// サーバにJSONを保存
Model.prototype.save = function () {
	var self = this;

	var method = self.isInitial() ? 'POST' : 'PUT';
	var url    = self.isInitial() ? api_url : api_url + "/" + self.id();

	return m.request({method: method, url: url, data: {
		name:        self.name(),
		admin:       self.admin.name(),
		start_date:  self.start_date(),
		capacity:    Number(self.capacity()), // int
		place:       self.place.name(),
		image:       self.image(),
		description: self.description(),
	}})
	.then(function(res) {
		self.isInitial(false); // サーバーにレコードが存在する
		// 生成されたイベントID
		return self.id() || res.id;
	});
};

// サーバからJSONを破棄
Model.prototype.destroy = function () {
	return m.request({
		method: "DELETE",
		url: api_url + "/" + this.id(),
		data: {}
	})
	.then(function(model) {
		model.isInitial(true); // サーバーにレコードが存在しない
	});

};

module.exports = Model;

File: resources/event.go
package resources

import (
	"github.com/labstack/echo"
	"github.com/syo-sa1982/GoNTAkun/model"
	"log"
	"net/http"
	"strconv"
	"time"
	"github.com/syo-sa1982/GoNTAkun/services"
)

func (resource *Resource) GetEvents() echo.HandlerFunc {

	return func(c echo.Context) error {

		var (
			db         = resource.DB
			events     = []model.Event{}
			events_res = []EventResponse{}
			viewCount  = 10
			current    int
			prev_id    int
			next_id    int
		)
		db = resource.SetDBConnection()
		defer db.Close()

		log.Println("p is ....")
		log.Println(c.QueryParam("p"))
		if c.QueryParam("p") != "" {
			current, _ = strconv.Atoi(c.QueryParam("p"))
		} else {
			current = 1
		}
		prev_id = current - 1
		next_id = current + 1

		db.Model(events).Offset((current - 1) * 3).Limit(viewCount).Find(&events)

		for i, s := range events {
			db.Model(&s).Related(&events[i].Members, "EventID")
		}

		events_res = resource.getEventsResponse(events)
		response := EventListAPI{prev_id, next_id, events_res}

		api := APIFormat{"success", 1, 0, response}

		return c.JSON(http.StatusOK, &api)
	}
}

func (resource *Resource) GetEvent() echo.HandlerFunc {

	return func(c echo.Context) error {
		log.Println("Start GetEvent")
		var (
			db        = resource.DB
			event     = model.Event{}
			event_res = EventResponse{}
			admin     = model.Member{}
			members   = []model.Member{}
			comments  = []model.Comment{}
		)
		db = resource.SetDBConnection()
		defer db.Close()

		db.Where("id = ?", c.Param("id")).Find(&event)
		db.Model(&event).Related(&admin, "Admin").Related(&members, "Members").Related(&comments, "Comments")

		event.Admin = admin
		event.Members = members
		event.Comments = comments

		event_res = resource.getEventResponse(event)

		api := APIFormat{"success", 1, 0, event_res}
		return c.JSON(http.StatusOK, &api)
	}
}

func (resource *Resource) CreateEvent() echo.HandlerFunc {

	return func(c echo.Context) error {
		log.Println("Start CreateEvent")
		var (
			db = resource.DB
			googleAccount = model.GoogleAccount{}
			account = model.Account{}
		)
		db = resource.SetDBConnection()
		defer db.Close()

		// クッキーからIDを取得する
		strId := ""
		id , _:= c.Cookie("id")
		if id != nil && id.Value() != "" {
			// IDを複合する
			strId = services.DecrypterBase64(id.Value())
		}

		if strId == "" {
			// TODO: エラー処理
			//return
		} else {
			// アカウントテーブルを取得
			db.Model(account).Where("id = ?", strId, ).Find(&account)

			// googleアカウントテーブルを取得
			db.Model(googleAccount).Where("g_id = ?", account.GID, ).Find(&googleAccount)
		}

		u := new(EventRequest)

		if err := c.Bind(u); err != nil {
			return err
		}

		t, _ := time.Parse(date_time_format, u.StartDate)

		event := model.Event{
			Name:        u.Name,
			Image:       u.Image,
			StartDate:   t,
			Capacity:    u.Capacity,
			Admin:       model.Member{AccountId: account.ID, Name: googleAccount.Name, Picture: googleAccount.Picture, AdminStatus: 1},
			Place:       u.Place,
			Description: u.Description,
			Comments:    []model.Comment{},
		}
		db.Create(&event)

		responseApi := map[string]int{"id": event.ID}

		api := APIFormat{"success", 1, 0, responseApi}
		return c.JSON(http.StatusOK, &api)
	}
}

func (resource *Resource) UpdateEvent() echo.HandlerFunc {

	return func(c echo.Context) error {

		var (
			db    = resource.DB
			event = model.Event{}
		)
		db = resource.SetDBConnection()
		defer db.Close()

		u := new(EventRequest)

		if err := c.Bind(u); err != nil {
			return err
		}

		db.Model(event).Where(c.Param("id")).Find(&event)

		t, _ := time.Parse(date_format, u.StartDate)

		event.Name = u.Name
		event.StartDate = t
		event.Capacity = u.Capacity
		event.Place = u.Place
		event.Image = u.Image
		event.Description = u.Description

		db.Model(event).Update(&event)

		api := APIFormat{"success", 1, 0, event}

		return c.JSON(http.StatusOK, &api)
	}
}

func (resource *Resource) DeleteEvent() echo.HandlerFunc {

	return func(c echo.Context) error {
		var (
			db    = resource.DB
			event = model.Event{}
		)
		db = resource.SetDBConnection()
		defer db.Close()

		responseApi := map[string]string{"ID": c.Param("id")}

		db.Model(event).Where(c.Param("id")).Delete(&event)

		api := APIFormat{"success", 1, 0, responseApi}
		return c.JSON(http.StatusOK, &api)
	}
}
